Changes in module 'nf-core/genomad/download'
--- modules/nf-core/genomad/download/main.nf
+++ modules/nf-core/genomad/download/main.nf
@@ -4,7 +4,7 @@
     conda "bioconda::genomad=1.5.2"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/genomad:1.5.2--pyhdfd78af_0':
-        'biocontainers/genomad:1.5.2--pyhdfd78af_0' }"
+        'quay.io/biocontainers/genomad:1.5.2--pyhdfd78af_0' }"
 
     output:
     path "genomad_db/"  , emit: genomad_db

************************************************************
