Changes in module 'nf-core/genomad/endtoend'
--- modules/nf-core/genomad/endtoend/main.nf
+++ modules/nf-core/genomad/endtoend/main.nf
@@ -5,7 +5,7 @@
     conda "bioconda::genomad=1.5.2"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/genomad:1.5.2--pyhdfd78af_0':
-        'biocontainers/genomad:1.5.2--pyhdfd78af_0' }"
+        'quay.io/biocontainers/genomad:1.5.2--pyhdfd78af_0' }"
 
     input:
     tuple val(meta) , path(fasta)

************************************************************
